javaparser-core/src/main/java/com/github/javaparser/HasParentNode.java

```diff
@@ -0,0 +1,49 @@
+package com.github.javaparser;
+
+import com.github.javaparser.ast.Node;
+import com.github.javaparser.ast.NodeList;
+
+import java.util.List;
+
+/**
+ * An object that has a parent node.
+ */
+public interface HasParentNode<T> {
+    // TODO nullable
+    Node getParentNode();
+
+    T setParentNode(Node parentNode);
+
+    /**
+     * "this" for everything except NodeLists. NodeLists use their parent as their childrens parent.
+     */
+    Node getParentNodeForChildren();
+
+    @SuppressWarnings("unchecked")
+    default <N> N getParentNodeOfType(Class<N> classType) {
+        Node parent = getParentNode();
+        while (parent != null) {
+            if (classType.isAssignableFrom(parent.getClass()))
+                return (N) parent;
```

Comment:
I'll pull this stuff, can you scan the codebase for these casts? They happen quite often.